feat(contribution):[TRACEFOSS-XXX] Added pre-commit hook to enforce c…

ds-mwesener opened the pull request:
…ommit messages have correct format.

ds-mwesener commented:
Hi @ds-ext-jserwa thanks for asking. Basically it will only enforce local commit message to have that format. But usually when you click on squash you will use that commit message either way. So if you dont change it manually in the git ui you will have unified commit messages.

dblankevia commented on dev/commit-msg:
```diff
@@ -0,0 +1,23 @@
+url="$(git remote get-url origin | sed 's/\.git/#commit-messages/')"
+
+# check semantic versioning scheme
+if ! echo "$commitTitle" | grep -qE '^(build|chore|ci|docs|feat|fix|perf|refactor|revert|style|test)\([a-z0-9/-]+\)\:\[[A-Z0-9-]+\].*$'; then
```
We wanted to shorten that list a bit. I suggest `chore|feat|fix`. All others are either 'chore' or 'feat' IMHO.

Additionally, Semantic Versioning and commit comments are totally unrelated. We should not use the term 'Semantic Versioning' here.

The regex also does not allow a space after the ':'.

Do we need the square brackets around the issue number? The commit message is already very long with all the boilerplate.

dblankevia replied:
I just saw the reference to https://github.com/hazcod/semantic-commit-hook. See also my comment to `CONTRIBUTING.md`.

dblankevia commented on CONTRIBUTING.md:
````diff
@@ -52,6 +52,22 @@ fulfills the DCO's requirement that you sign-off on your contributions.
+cp dev/commit-msg .git/hooks/commit-msg && chmod 500 .git/hooks/commit-msg
+```
+
+For further information please see https://github.com/hazcod/semantic-commit-hook
````
Our guidelines say, that we write the change log manually:
https://confluence.catena-x.net/display/BDPQ/2.+(TRF)+Development+Process
https://keepachangelog.com/en/1.0.0/

So we're not using commit comments to generate the commit log and neither to automatically calculate a version number. We should therefore not have this reference here.

ds-mwesener commented:
Hi I have changed implementation of regex to chore|feat|fix - also removed the must have brackets [] and allowed an optional space after colon. Additionally changed contribution.md

evia-fbedarf commented on CONTRIBUTING.md:
```diff
@@ -52,6 +52,20 @@ fulfills the DCO's requirement that you sign-off on your contributions.
 
+### Commit messages
+The commit messages have to match a pattern in the form of:
+``< type >(optional scope):[<Ticket_ID>] < description >``
```
Space missing after the colon

evia-fbedarf commented on CONTRIBUTING.md:
```diff
@@ -52,6 +52,20 @@ fulfills the DCO's requirement that you sign-off on your contributions.
+Example:
+``feat(users): TRACEFOSS-XXX Description``
+
+Detailed pattern can be found here: [commit-msg](./dev/commit-msg)
```
Detailed > The detailed

ds-mwesener replied:
Done.

evia-fbedarf commented on dev/commit-msg:
```diff
@@ -0,0 +1,23 @@
+url="$(git remote get-url origin | sed 's/\.git/#commit-messages/')"
+
+# check semantic versioning scheme
+if ! echo "$commitTitle" | grep -qE '^(chore|fix|feat)\([a-z0-9\/-]+\)\s?:\s?[A-Z0-9-]+.*$'; then
```
We discussed not using abbreviations, so let us use "feature" instead of "feat".

We need to clarify if we want to add these two sections as well:
- revert
- test
Referencing this ticket: https://jira.catena-x.net/browse/TRACEFOSS-1007

dblankevia replied:
I propose to start without them. Revert is usually a fix and test is chore or feature?

ds-mwesener replied:
I agree with @dblankevia I dont like to have to many types - because basically the dev takes the decision which message he writes - as this sometimes can be hard to classify, this leads to different message usages.

evia-fbedarf commented on dev/commit-msg:
```diff
@@ -0,0 +1,23 @@
+url="$(git remote get-url origin | sed 's/\.git/#commit-messages/')"
+
+# check semantic versioning scheme
+if ! echo "$commitTitle" | grep -qE '^(chore|fix|feat)\([a-z0-9\/-]+\)\s?:\s?[A-Z0-9-]+.*$'; then
```
The scope should be optional fix(optional scope):  TRACEFOSS-xxx

evia-fbedarf commented on dev/commit-msg:
```diff
@@ -0,0 +1,23 @@
+url="$(git remote get-url origin | sed 's/\.git/#commit-messages/')"
+
+# check semantic versioning scheme
+if ! echo "$commitTitle" | grep -qE '^(chore|fix|feat)\([a-z0-9\/-]+\)\s?:\s?[A-Z0-9-]+.*$'; then
```
Could you add a validation check so that the ticket ID is checked for semantics?

Maybe we should allow "TRACEFOS-xxx" if there is no ticket or something.

ds-mwesener replied:
Hi is already allowed. E.g. feat(users): TRACEFOSS-XXX Description

evia-fbedarf reviewed (approved):
LGTM
